src/Interfaces.jl

```diff
@@ -83,16 +94,24 @@ function toimplements!(T, arg::Expr, shouldthrow::Bool=true)
     elseif arg.head == :(::)
         # required method definition and required return type
         nm, args = methodparts(T, arg.args[1])
-        __RT__ = arg.args[2]
         sym = gensym()
-        return quote
-            check = $(requiredmethod(T, nm, args, shouldthrow))
-            $sym = Interfaces.returntype($nm, $args)
-            # @show $sym, $nm, $args, Interfaces.isinterfacetype($__RT__)
-            check |= Interfaces.isinterfacetype($__RT__) ?
-                Interfaces.implements($sym, $__RT__) : $sym <: $__RT__
-            check || ($shouldthrow && Interfaces.invalidreturntype($nm, $args, $sym, $__RT__))
+        __RT__ = arg.args[2]
+        return requiredreturn(T, nm, args, shouldthrow, sym, __RT__)
+    elseif arg.head == :->
+        # required method definition with captured return type name
+        nm, args = methodparts(T, arg.args[1])
+        blockargs = arg.args[2].args
+        length(blockargs) == 1 || invalidnamedreturn()
+        if blockargs[1] isa Symbol
+            sym = blockargs[1]
+            __RT__ = Any
+        elseif blockargs[1].head == :(::)
+            sym = blockargs[1].args[1]
+            __RT__ = blockargs[1].args[2]
```

Comment:
ah, we need to handle (gracefully disallow) `foo(::Blah) -> ::Iterable` i.e. using the `->` syntax for _unnamed_ return type annotation (which is instead to be written as `foo(::Blah)::Iterable`)

Reply:
In my `foo() -> T || foo() -> x::T` proposal, we would allow a single Symbol (type), or `Expr(:(::), arg1, arg2)` form which would be `name::Type`, so we could error in any other cases by saying it's an invalid return type specification.